src/Database/LSMTree/Internal/WriteBuffer.hs

```diff
@@ -53,55 +62,100 @@ import           Prelude hiding (lookup)
+
+sizeInBytes :: WriteBuffer -> Int
+sizeInBytes = writeBufferSizeBytes
+
+fromMap ::
+     Map SerialisedKey (Entry SerialisedValue SerialisedBlob)
+  -> WriteBuffer
+fromMap m = WB m (Map.foldlWithKey' (\s k e -> s + sizeofKOp k e) 0 m)
+  where
+    sizeofKOp k e = sizeofKey k + onValue 0 sizeofValue e
+
+fromList ::
+     (SerialisedValue -> SerialisedValue -> SerialisedValue) -- ^ merge function
+  -> [(SerialisedKey, Entry SerialisedValue SerialisedBlob)]
+  -> WriteBuffer
+fromList f es = fromMap (Map.fromListWith (combine f) es)
 
 -- | \( O(n) \)
-content :: WriteBuffer ->
-           [(SerialisedKey, Entry SerialisedValue SerialisedBlob)]
-content (WB m) = Map.assocs m
+toList :: WriteBuffer -> [(SerialisedKey, Entry SerialisedValue SerialisedBlob)]
+toList (WB m _) = Map.assocs m
 
 {-------------------------------------------------------------------------------
   Updates
 -------------------------------------------------------------------------------}
 
+-- TODO(optimisation): make sure the core looks reasonable, e.g. not allocating
+-- 'WB' for every entry.
+addEntries ::
+     Foldable f
+  => (SerialisedValue -> SerialisedValue -> SerialisedValue) -- ^ merge function
+  -> f (SerialisedKey, Entry SerialisedValue SerialisedBlob)
+  -> WriteBuffer
+  -> WriteBuffer
+addEntries f es wb = foldl' (flip (uncurry (addEntry f))) wb (Foldable.toList es)
+
 addEntry ::
      (SerialisedValue -> SerialisedValue -> SerialisedValue) -- ^ merge function
   -> SerialisedKey
   -> Entry SerialisedValue SerialisedBlob
   -> WriteBuffer
   -> WriteBuffer
-addEntry f k e (WB wb) =
-    WB (Map.insertWith (combine f) k e wb)
+addEntry f k e (WB wb s) =
+    let (!wb', !s') = runState (insert k wb) s
+    in WB wb' s'
+  where
+    -- TODO: this seems inelegant, but we want to avoid traversing the Map twice
+    insert = Map.alterF $ (fmap Just .) $ \case
+        Nothing -> do
+          modify' (+ (sizeofKey k + sizeofEntry e))
+          return e
+        Just old -> do
+          let !new = combine f e old
+          -- don't count key (it was already there), only count value difference
+          modify' (+ (sizeofEntry new - sizeofEntry old))
+          return new
+    sizeofEntry = onValue 0 sizeofValue
 
 addEntryMonoidal ::
      (SerialisedValue -> SerialisedValue -> SerialisedValue) -- ^ merge function
-  -> SerialisedKey -> Monoidal.Update SerialisedValue -> WriteBuffer -> WriteBuffer
-addEntryMonoidal f k e (WB wb) =
-    WB (Map.insertWith (combine f) k (updateToEntryMonoidal e) wb)
+  -> SerialisedKey
+  -> Monoidal.Update SerialisedValue
+  -> WriteBuffer
+  -> WriteBuffer
+addEntryMonoidal f k = addEntry f k . updateToEntryMonoidal
 
 addEntryNormal ::
      SerialisedKey
   -> Normal.Update SerialisedValue SerialisedBlob
   -> WriteBuffer
   -> WriteBuffer
-addEntryNormal k e (WB wb) =
-    WB (Map.insert k (updateToEntryNormal e) wb)
+addEntryNormal k = addEntry const k . updateToEntryNormal
```

Comment:
I'm not sure we actually use these functions. They could be cleaned up (but doesn't have to be in this PR)